substratevm/DEBUGINFO.md

```diff
@@ -25,50 +36,119 @@ compiled method.
+source file cached as sources/src/org/my/foo/Foo.java).
+
+The lookup scheme for cached pplication sources varies depending upon
+what is found in each classpath entry. Given a jar file entry like
+/path/to/foo.jar, the corresponding jar /path/to/foo-sources.jar is
+considered as a candidate zip file system from which source files may
+be extracted. When the entry specifies a dir like /path/to/bar/classes
+or /path/to/bar/target/classes then directory /path/to/bar/src is
+considered as a candidate. Finally, the current directory in whcih the
+native image program is being run is also considered as a candidate.
+
+These lookup strategies are only provisional and may need extending in
+future. Note however that it is possible to make missing sources
+available by other means. One option is to unzip extra app source jars
+or copying extra app source trees into the cache. Another is to
+configure extra source search paths (see below).
+
+Configuring source paths in gdb
+-------------------------------
 
-    (gdb) set directories /home/adinn/hello/src:/home/adinn/graal/sdk/src/org/graalvm.word/src:/home/adinn/graal/sdk/src/org.graalvm.options/src:...
-
-The argument is a comma separated list of source roots. It needs to
-identify:
+In order for gdb to be able to locate the source files for your app
+classes, Graal classes and JDK runtime classes you need to provide gdb
+with a list of source root dirs using the 'set directories' command:
 
-  - sources for your app
-  - sources under the Graal sdk, compiler, substratevm and truffle trees
-  - sources in the JDK src.zip file
+    (gdb) set directories /path/to/sources/jdk:/path/to/sources/graal:/path/to/sources/src
 
-Needless to say the list for Graal is long and complex. Also, the JDK
-sources are in a zip file and gdb does not understand zip sources. So
-you need to extract the JDK sources as a preparatory step,
+Directory .../sources/jdk should contain source files for all JDK runtime
+classes referenced from debug records.
 
-You can use shell script write_gdbsourcepath (added to Graal dir
-substratevm) to auto-generate settings for the GRaal and JDK sources.
+Directory .../sources/graal should contain source files for all GraalVM
+classes referenced from debug records. Note that the current
+implementation does not yet find some sources for the GraalVM JIT
+compiler in the org.graalvm.compiler* package subspace.
 
-    $ bash write_gdbsourcepath
+Directory .../sources/src should contain source files for all
+application classes referenced from debug records, assuming they can
+be located using the lookup strategy described above.
 
-It creates a local file .gdbsourcepath which sets the relevant
-directories. Before running it you can set two env vars to tell it
-where to locate the source trees it needs to include:
+You can supplement the files cached in sources/src by unzipping
+application source jars or copying application source trees into the
```

Comment:
That's cumbersome. Since you already have the infrastructure to copy sources from zip or dir for your automatic lookup strategy adding a hosted option that allows user to add additional lookup roots would be the most natural way to allow users to supplement source lookup.

Reply:
Sure I will I add something along the lines of CLibraryPath or LinkerRPath i.e. on the command line the user specifies a comma-separated list of dirs

    -X:DebugInfoSourceSearchPath=/path/to/dir1,/path/to/jar2.jar,/path/to/jar3.jar

which gets installed as a HostedOptionKey<String[]>

Reply:
I have added this in the latest push. As per the other paths the argument can be a single dir/jar/zip or a comma separated list and the option can appear multiple times if desired. Users call Options.flatten to squeeze to the comma separators. I still nee dto document this in DEBUGINFO.md

Reply:
Ok, DEBUGINFO.md now documents how to specify search paths.